R/crosswalk.R

```diff
@@ -21,154 +24,201 @@
-  y <- jsonld::jsonld_expand(y)
-  y <- jsonld::jsonld_compact(y, context = codemeta_context)
-  y
-}
+crosswalk_transform <- function(
+  x, crosswalk_context = NULL, codemeta_context = NULL
+) {
 
+  codemeta_context <- default_context_if_null(codemeta_context)
 
+  add_context(x, context = crosswalk_context) %>%
+    jsonlite::toJSON(auto_unbox = TRUE, pretty = TRUE) %>%
+    jsonld::jsonld_expand() %>%
+    jsonld::jsonld_compact(context = codemeta_context)
+}
 
 #' drop_context
 #'
 #' drop context element from json list or json string
-#' @param x a JSON list (from read_json / fromJSON) or json object (from toJSON)
-#' @param json_output logical, should output be a json object or a list?
+#'
+#' @inheritParams set_context
 #' @return a list or json object with the "@context" element removed
 #' @importFrom jsonlite toJSON fromJSON
 #' @export
-drop_context <-function(x, json_output = FALSE){
-  if(is(x, "json") || is.character(x)){
-    x <- fromJSON(x)
-    json_output <- TRUE
-  }
-  x$`@context` <- NULL
-
+#'
+drop_context <-function(x, json_output = FALSE) {
 
-  if(json_output){
-    x <- toJSON(x, auto_unbox = TRUE, pretty = TRUE)
-  }
-  x
+  # Remove context by setting the corresponding list element to NULL
+  set_context(x, NULL, json_output)
 }
 
 #' add_context
 #'
 #' add context element to json list or json string
 #'
-#' @param x a JSON list (from read_json / fromJSON) or json object (from toJSON)
+#' @inheritParams set_context
 #' @param context context to be added, in same format as x
-#' @param json_output logical, should output be a json object or a list?
 #' @return a list or json object with "@context" element added
 #' @export
-add_context <- function(x, context, json_output = FALSE){
-  if(is(x, "json") || is.character(x)){
-    x <- fromJSON(x)
-    json_output <- TRUE
-  }
-  if(is(context, "json") || is.character(context)){
-    context <- fromJSON(context)
-  }
+#'
+add_context <- function(x, context, json_output = FALSE) {
 
-  ## make sure context doesn't already have a "@context" property
-  x$`@context` <- context
+  new_context <- from_json_if(is_json_or_character(context), context)
 
-  if(json_output){
-   x <- toJSON(x, auto_unbox = TRUE, pretty = TRUE)
-  }
-  x
+  set_context(x, new_context, json_output)
 }
 
+#' set_context
+#'
+#' set context element in json list or json string
+#'
+#' @param x a JSON list (from read_json / fromJSON) or json object (from toJSON)
+#' @param new_context new value for the context element
+#' @param json_output logical, should output be a json object or a list?
+#' @return a list or json object with "@context" element set to \code{context}
+#'
+set_context <- function(x, new_context, json_output = FALSE) {
+
+  # is x of class "json" or character?
+  json_input <- is_json_or_character(x)
+
+  # if yes, call fromJSON() on x
+  x <- from_json_if(json_input, x)
+
+  ## TODO: make sure context doesn't already have a "@context" property
+  x$`@context` <- new_context
```

Comment:
This is fine being left as is, but just curious while we're doing this -- do you have any preference for 

```
x$`@context`
```

vs 

```
x[["@context"]]
```

In general I agree `$` subsetting is cleaner / more concise, but backtick escapes always feel irregular to me.

Reply:
Very good.  Let's change this convention then as well to avoid using backticks.

Reply:
What does the `@` character indicate? Does it indicate that a field is required? For that case I was thinking about a function `set_required_field`:
```
x <- set_required_field(context = new_context)
```
But maybe this is "overkill"?

Reply:
Backticks replaced, see commit 95c90cc